Initial implementation

rashmihunt opened the pull request:
Opening another PR since the first one got closed.. @raymondfeng PTAL

raymondfeng commented on index.js:
```diff
@@ -0,0 +1,47 @@
+ * @returns {String}
+ */
+exports.generateCode = function(version, modelName, operations) {
+  return getGenerator(spec).generateCodeForOperations(modelName, operations);
```
`spec` is not defined. Should the method signature be `exports.generateCode = function(spec, modelName, operations)`?

rashmihunt commented:
@raymondfeng I am fixing eslint issues. Meanwhile, you could run the mocha tests in webstorm so that you can see the output of code generation. Webstorm run doesn't show the wslint issues.

rmg commented on index.js:
```diff
@@ -0,0 +1,47 @@
+// Copyright IBM Corp. 2014,2016. All Rights Reserved.
```
Please re-run `apict copyright` to fix the copyright headers (there should only be a single year for this file, for example).

rmg commented on test/results/foo_model_nested_ref.json:
```diff
@@ -0,0 +1,51 @@
+      }
+    }
+  }
+}
```
Missing terminating newline. You may need to configure your editor to do this properly.

rmg commented on test/results/oneway_model.json:
```diff
@@ -0,0 +1,10 @@
+      }
+    }
+  }
+}
```
Missing terminating newline. You may need to configure your editor to do this properly.

rmg commented on test/results/recursive_model.json:
```diff
@@ -0,0 +1,41 @@
+      }
+    }
+  }
+}
```
Missing terminating newline. You may need to configure your editor to do this properly.

rashmihunt commented:
@slnode test please

raymondfeng commented on test/test.js:
```diff
@@ -81,10 +80,10 @@ describe('Generate APIs and models with WSDLs containing ', function() {
 
           var code = helper.generateRemoteMethods(apiData);
-          //console.log(code);
+          // console.log(code);
```
Please remove the commented out code.

rashmihunt replied:
I purposely left it commented out console.log statements for easy debugging purpose for future failures if any... I can remove it.

rmg commented on .npmignore:
```diff
@@ -0,0 +1,20 @@
+lib-cov
```
The `.npmignore` file should probably be deleted.

rashmihunt replied:
hmm I thought I did delete this file.. I will do it.

rashmihunt commented:
@rmg  I have incorporated all your comments. PTAL
@raymondfeng PTAL  1) incorporated your comments 
2) Using chai for JSON comparison 
3) Fixed eslint issues 
4) Fixed package.json with latest version of dependecies 
5) Removed console.log statements

rmg commented on lib/codegen/model.ejs:
```diff
@@ -0,0 +1,87 @@
+        }
+        requestArgs = param.arg + ": " + param.arg;
+    }
+    requestArgs = "{\n" + "             " + requestArgs + "\n        }";
```
Since this module required node v4 or newer you might want to consider using the new [template string literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) that were introduced in ES2015. In this case you would probably only be using them for simple multi-line strings with interpolation support:
```js
requestArgs = `{
             ${requestArgs}
        }`;
```

rmg commented on package.json:
```diff
@@ -27,10 +27,22 @@
+    "spawn-sync": "1.0.15",
+    "strong-globalize": "^2.6.0",
+    "strong-soap": "^1.0.1",
+    "underscore.string": "^3.3.4"
```
What is this dependency used for? I don't see any references to it.

rashmihunt replied:
strong soap is used in model.ejs --> var soap = require('strong-soap').soap;

rashmihunt replied:
"underscore.string": "^3.3.4" used in generator-soap -->      op.operationId = _s.camelize(
                op.operationId.split(/[\.\s\/-:\*]+/).join('_'));

rmg commented on package.json:
```diff
@@ -27,10 +27,22 @@
+    "debug": "^2.2.0",
+    "ejs": "^2.5.5",
+    "lodash": "^4.17.4",
+    "spawn-sync": "1.0.15",
```
This dependency appears unused. It is also redundant since Node v4 includes `child_process.spawnSync`.

rmg commented on example/example.js:
```diff
@@ -0,0 +1,43 @@
+var helper = require('../index');
+var soap = require('strong-soap').soap;
+var WSDL = soap.WSDL;
+var curl = require('curl');
```
I don't see where `curl` is actually used.

rmg commented on package.json:
```diff
@@ -27,10 +27,22 @@
+    "curl": "^0.1.4",
+    "debug": "^2.2.0",
+    "ejs": "^2.5.5",
+    "lodash": "^4.17.4",
```
This appears to be unused

rashmihunt replied:
ejs is used in code generator-soap.js ->  var code = ejs.render(template,  ...

rashmihunt replied:
lodash is used in generator-soap.js --> var copy = _cloneDeep(op || {});

rashmihunt replied:
@rmg PTAL

rmg commented on lib/codegen/generator-soap.js:
```diff
@@ -0,0 +1,429 @@
+
+            // Camelize the operation id
+      op.operationId = op.operationId.replace(/{(([^{}])+)}/g, '_$1');
+      op.operationId = _s.camelize(
```
It looks like this one call is the only reason `underscore.string` is used at all. Does it do something that `_.camelCase` from lodash doesn't do?

rmg commented:
My mistake, it looks like GitHub was suppressing that part of the diff for some reason.

rmg commented on package.json:
```diff
@@ -27,10 +27,17 @@
     "eslint": "^3.12.2",
-    "eslint-config-loopback": "^7.0.0"
+    "eslint-config-loopback": "^8.0.0",
+    "loopback": "^3.0.0",
```
This package doesn't actually use loopback, as far as I can tell. I was able to remove this line and `git clean -fxd && npm install && npm test` still passed.

rashmihunt replied:
@rmg I am hesitant to remove it for now.. I will do end to end test (not just module level test) and see if it works using 'lb soap' command and REST APIs can be invoked from APIExplorer. If this works then, I will remove "loopback": "^3.0.0", in another PR.. I have another work item to add more test cases to this module in next sprint anyway.

rmg replied:
Normally I'd ask this to be removed until then, but you've already put up with enough of my demands I'll let this one slide :-P
